internal/provider/gce/environ_broker.go

```diff
@@ -166,6 +167,48 @@ func formatMachineType(zone, name string) string {
 	return fmt.Sprintf("zones/%s/machineTypes/%s", zone, name)
 }
 
+func (env *environ) serviceAccount(ctx context.ProviderCallContext, args environs.StartInstanceParams) (string, error) {
+	var serviceAccount string
+
+	// For controllers, the service account can come from the credential.
+	if args.InstanceConfig.IsController() {
+		if env.cloud.Credential.AuthType() == cloud.ServiceAccountAuthType {
+			serviceAccount = env.cloud.Credential.Attributes()[credServiceAccount]
+			if serviceAccount != "" {
+				logger.Debugf("using credential service account: %s", serviceAccount)
+			}
+		} else if args.InstanceConfig.Bootstrap != nil && args.InstanceConfig.Bootstrap.BootstrapMachineConstraints.HasInstanceRole() {
+			serviceAccount = *args.InstanceConfig.Bootstrap.BootstrapMachineConstraints.InstanceRole
+			if serviceAccount != "" {
+				logger.Debugf("using bootstrap service account: %s", serviceAccount)
+			}
+		}
+	}
+	if serviceAccount != "" {
+		return serviceAccount, nil
+	}
+
+	// Next use the constraint value if supplied.
+	if args.Constraints.HasInstanceRole() {
+		serviceAccount = *args.Constraints.InstanceRole
```

Comment:
This is not fully clear to me. We can pass the service account value through the instance role constraint, but then why did you have to add a new auth type in https://github.com/juju/juju/pull/20585/files#diff-b5dd98f1146a7c49a0ad8dfd82ee36c9104e300eb25efbf710af3442d6268043R70-R74 instead of using `InstanceRoleAuthType`?

Reply:
Because it's not instance role. Sadly when instance role was done, ec2 specific naming was used for the constraint attribute and we need to retain it for compatibility. The credential auth type  for gce is service account. For ec2 it's instance role. And for Azure, where we also support in cloud auth, it's managed identity. Each cloud has its own auth type for this since the implementations are cloud specific.